zcash_client_sqlite/src/wallet/init.rs

```diff
@@ -40,106 +431,40 @@ use {
-            FOREIGN KEY (block) REFERENCES blocks(height),
-            CONSTRAINT witness_height UNIQUE (note, block)
-        )",
-        NO_PARAMS,
-    )?;
-    wdb.conn.execute(
-        "CREATE TABLE IF NOT EXISTS sent_notes (
-            id_note INTEGER PRIMARY KEY,
-            tx INTEGER NOT NULL,
-            output_pool INTEGER NOT NULL,
-            output_index INTEGER NOT NULL,
-            from_account INTEGER NOT NULL,
-            address TEXT NOT NULL,
-            value INTEGER NOT NULL,
-            memo BLOB,
-            FOREIGN KEY (tx) REFERENCES transactions(id_tx),
-            FOREIGN KEY (from_account) REFERENCES accounts(account),
-            CONSTRAINT tx_output UNIQUE (tx, output_pool, output_index)
-        )",
-        NO_PARAMS,
-    )?;
-    wdb.conn.execute(
-        "CREATE TABLE IF NOT EXISTS utxos (
-            id_utxo INTEGER PRIMARY KEY,
-            address TEXT NOT NULL, 
-            prevout_txid BLOB NOT NULL, 
-            prevout_idx INTEGER NOT NULL, 
-            script BLOB NOT NULL, 
-            value_zat INTEGER NOT NULL, 
-            height INTEGER NOT NULL,
-            spent_in_tx INTEGER,
-            FOREIGN KEY (spent_in_tx) REFERENCES transactions(id_tx),
-            CONSTRAINT tx_outpoint UNIQUE (prevout_txid, prevout_idx)
-        )",
-        NO_PARAMS,
-    )?;
+// TODO: It would be possible to make the transition from providing transparent support to no
+// longer providing transparent support safe, by including a migration that verifies that no
+// unspent transparent outputs exist in the wallet at the time of upgrading to a version of
+// the library that does not support transparent use. It might be a good idea to add an explicit
+// check for unspent transparent outputs whenever running initialization with a version of the
+// library *not* compiled with the `transparent-inputs` feature flag, and fail if any are present.
+pub fn init_wallet_db<P: consensus::Parameters + 'static>(
+    wdb: &mut WalletDb<P>,
+    seed: Option<SecretVec<u8>>,
+) -> Result<(), MigratorError<WalletMigrationError>> {
+    wdb.conn
+        .execute("PRAGMA foreign_keys = OFF", NO_PARAMS)
+        .map_err(|e| MigratorError::Adapter(WalletMigrationError::from(e)))?;
+    let adapter = RusqliteAdapter::new(&mut wdb.conn, Some("schemer_migrations".to_string()));
+    adapter.init().expect("Migrations table setup succeeds.");
+
+    let mut migrator = Migrator::new(adapter);
+    let migration0 = Box::new(WalletMigration0 {});
+    let migration1 = Box::new(WalletMigration1 {});
+    let migration2 = Box::new(WalletMigration2 {
+        params: wdb.params.clone(),
+        seed,
+    });
+
+    migrator
+        .register_multiple(vec![migration0, migration1, migration2])
+        .expect("Wallet migration registration should have been successful.");
+    migrator.up(None)?;
+    wdb.conn
+        .execute("PRAGMA foreign_keys = ON", NO_PARAMS)
```

Comment:
This again doesn't enforce foreign_key constraints when creating the `sent_notes_new` table. I don't object, but why was this changed from a previous commitment that did?

Reply:
The pragmas only work per-connection, not within a transaction (as you identified in the docs) so we now drop constraints prior to all the migrations and reenable them all afterward.

Reply:
That makes sense as far as it goes, but how was the PR passing tests when the PRAGMA was inside the transaction?

Reply:
At the time, there was no data in the tables, so the foreign key constraints (at the row level) weren't being checked. I'm not sure why the DDL was succeeding though.